### PR TITLE
Implementation of `VFPINJ` curves in the `VFP` module. 

### PR DESCRIPTION
**Pull request by lindjoha**
Implementation of `VFPINJ` curves in the `VFP` module.

### CODE REVIEW

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 1]
```diff
@@ -49,43 +52,54 @@ export class VfpDataAccessor {
...
         if (vfpParam == VfpParam.THP) {
             return this._vfpTable.thp_values
-        } else if (vfpParam == VfpParam.WFR) {
-            return this._vfpTable.wfr_values
-        } else if (vfpParam == VfpParam.GFR) {
-            return this._vfpTable.gfr_values
-        } else if (vfpParam == VfpParam.ALQ) {
-            return this._vfpTable.alq_values
         }
+        if ("wfr_values" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:56]: Consider using suggested `isProdTable`-utlity for if-check

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 2]
```diff
@@ -49,43 +52,54 @@ export class VfpDataAccessor {
...
         if (vfpParam == VfpParam.THP) {
             return this._vfpTable.thp_unit
-        } else if (vfpParam == VfpParam.WFR) {
-            return this._vfpTable.wfr_unit
-        } else if (vfpParam == VfpParam.GFR) {
-            return this._vfpTable.gfr_unit
-        } else if (vfpParam == VfpParam.ALQ) {
-            return this._vfpTable.alq_unit
+        } 
+        if ("wfr_values" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:73]: Consider using suggested `isProdTable`-utlity for if-check

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 3]
```diff
@@ -49,43 +52,54 @@ export class VfpDataAccessor {
...
-        } else if (vfpParam == VfpParam.GFR) {
-            label = this._vfpTable.gfr_type
-        } else if (vfpParam == VfpParam.ALQ) {
-            if (this._vfpTable.alq_type === ALQ_api._) {
-                label = "ALQ"
-            } else {
-                label = "ALQ: " + this._vfpTable.alq_type
-            }
+        } 
+        if ("wfr_values" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:91]: Consider using suggested `isProdTable`-utlity for if-check

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 4]
```diff
@@ -95,35 +109,59 @@ export class VfpDataAccessor {
...
 
-    getBhpValues(thpIndex: number, wfrIndex: number, gfrIndex: number, alqIndex: number) : number[] {
-        const nbWfrValues = this._vfpTable.wfr_values.length
-        const nbGfrValues = this._vfpTable.gfr_values.length
-        const nbAlqValues = this._vfpTable.alq_values.length
-        const nbFlowRates = this._vfpTable.flow_rate_values.length
-        const startIndex = nbFlowRates*(nbAlqValues*(nbGfrValues*(nbWfrValues*thpIndex+wfrIndex)+gfrIndex)+alqIndex)
-        return this._vfpTable.bhp_values.slice(startIndex, startIndex+nbFlowRates)
+    getVfpProdBhpValues(thpIndex: number, wfrIndex: number, gfrIndex: number, alqIndex: number) : number[] {
+        if ("wfr_values" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:133]: Consider using suggested `isProdTable`-utlity for if-check.

I'd reverse the if-check to, s.t. error is thrown in if-statement and larger amount of code is outside if:

```
if (!isProdTable(this._vfpTable)){
  throw Error("The getVfpProdBhpValues function is not valid for VFPINJ tables.")
}

const nbWfrValues = this._vfpTable.wfr_values.length
const nbGfrValues = this._vfpTable.gfr_values.length
const nbAlqValues = this._vfpTable.alq_values.length
const nbFlowRates = this._vfpTable.flow_rate_values.length
const startIndex = nbFlowRates*(nbAlqValues*(nbGfrValues(nbWfrValues*thpIndex+wfrIndex)+gfrIndex)+alqIndex)
return this._vfpTable.bhp_values.slice(startIndex, startIndex+nbFlowRates)

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 5]
```diff
@@ -95,35 +109,59 @@ export class VfpDataAccessor {
...
-        return this._vfpTable.wfr_type
+        if ("wfr_type" in this._vfpTable) { // This means that it is a VFPPROD table
+            return this._vfpTable.wfr_type
+        }
+        throw Error("WFR type is not valid for VFPINJ tables.")
     }
 
     getGfrType(): GFR_api {
-        return this._vfpTable.gfr_type
+        if ("gfr_type" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:119]: Consider using suggested `isProdTable`-utlity for if-check.

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 6]
```diff
@@ -95,35 +109,59 @@ export class VfpDataAccessor {
...
+            return this._vfpTable.gfr_type
+        }
+        throw Error("GFR type is not valid for VFPINJ tables.")
     }
 
     getAlqType(): ALQ_api {
-        return this._vfpTable.alq_type
+        if ("alq_type" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:126]: Consider using suggested `isProdTable`-utlity for if-check.

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 7]
```diff
@@ -95,35 +109,59 @@ export class VfpDataAccessor {
     }
 
     getWfrType(): WFR_api {
-        return this._vfpTable.wfr_type
+        if ("wfr_type" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:112]: Consider using suggested `isProdTable`-utlity for if-check.

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 8]
```diff
@@ -95,35 +109,59 @@ export class VfpDataAccessor {
...
+        throw Error("The getVfpProdBhpValues function is not valid for VFPINJ tables.")
+    }
+
+    getVfpInjBhpValues(thpIndex: number) : number[] {
+        if (this._vfpTable.vfp_type === VfpType_api.VFPINJ) {
+            const nbFlowRates = this._vfpTable.flow_rate_values.length
+            const startIndex = nbFlowRates*thpIndex
+            return this._vfpTable.bhp_values.slice(startIndex, startIndex+nbFlowRates)
+        }
+        throw Error("The getVfpInjBhpValues function is not valid for VFPPROD tables.")
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:150]: Consider using suggested `isProdTable`-utlity for if-check.

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 9]
```diff
@@ -95,35 +109,59 @@ export class VfpDataAccessor {
...
         if (vfpParam == VfpParam.THP) {
             return this._vfpTable.thp_values.length
-        } else if (vfpParam == VfpParam.WFR) {
-            return this._vfpTable.wfr_values.length
-        } else if (vfpParam == VfpParam.GFR) {
-            return this._vfpTable.gfr_values.length
-        } else if (vfpParam == VfpParam.ALQ) {
-            return this._vfpTable.alq_values.length
+        } 
+        if ("wfr_values" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:157]: Consider using suggested `isProdTable`-utlity for if-check.

File: frontend/src/modules/Vfp/utils/VfpPlotBuilder.ts [hunk 10]
```diff
@@ -97,14 +97,91 @@ export class VfpPlotBuilder {
...
-    private getSingleVfpTrace(thpIndex: number, wfrIndex: number, gfrIndex: number, alqIndex: number, pressureOption: PressureOption, color: string) : Partial<PlotData> { 
+
+    makeVfpInjTraces(
+        selectedThpIndices: number[] | null,
+        pressureOption: PressureOption,
+    ) : Partial<PlotData>[] {
+
+        const data: Partial<PlotData>[] = [];
+
+        if (selectedThpIndices == null) {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpPlotBuilder.ts:108]: Use strict equality ===

File: frontend/src/modules/Vfp/view.tsx [hunk 11]
```diff
@@ -46,14 +48,23 @@ export function View({ viewContext, workbenchSettings }: ModuleViewProps<Interfa
...
-        const data = vfpPlotBuilder.makeTraces(
-            selectedThpIndices,
-            selectedWfrIndices,
-            selectedGfrIndices,
-            selectedAlqIndices,
-            selectedPressureOption,
-            selectedColorBy,
-        );
+        let data: Partial<PlotData>[] = [];
+        if (vfpTable.vfp_type == VfpType_api.VFPPROD) {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/view.tsx:52]: Use strict equality ===

File: frontend/src/modules/Vfp/view.tsx [hunk 12]
```diff
@@ -46,14 +48,23 @@ export function View({ viewContext, workbenchSettings }: ModuleViewProps<Interfa
...
+            data = vfpPlotBuilder.makeVfpProdTraces(
+                selectedThpIndices,
+                selectedWfrIndices,
+                selectedGfrIndices,
+                selectedAlqIndices,
+                selectedPressureOption,
+                selectedColorBy,
+            )
+        } else if (vfpTable.vfp_type == VfpType_api.VFPINJ) {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/view.tsx:61]: Use strict equality ===

File: backend_py/primary/primary/services/sumo_access/vfp_access.py [hunk 13]
```diff
@@ -76,47 +77,75 @@ async def get_vfp_table_from_tagname_as_pyarrow(self, tagname: str, realization:
...
+                tab_type=tab_type,
+                thp_values=thp_values,
+                flow_rate_values=flow_rate_values,
+                bhp_values=[val for sublist in np.array(pa_table.columns).tolist() for val in sublist],
+                flow_rate_unit=VFP_UNITS[unit_type][VfpParam.FLOWRATE][flow_rate_type],
+                thp_unit=VFP_UNITS[unit_type][VfpParam.THP][thp_type],
+                bhp_unit=VFP_UNITS[unit_type][VfpParam.THP][thp_type],
+            )
+
+        if vfp_type == VfpType.VFPPROD:
```
Comment (jorgenherje) [inline backend_py/primary/primary/services/sumo_access/vfp_access.py:116]: Use strict equality ===

File: frontend/src/modules/Vfp/settings/atoms/derivedAtoms.ts [hunk 14]
```diff
@@ -143,10 +158,14 @@ export const selectedPressureOptionAtom = atom<PressureOption>((get) => {
 });
 
 export const selectedColorByAtom = atom<VfpParam>((get) => {
+    const vfpTable = get(vfpTableQueryAtom).data;
     const userSelectedColorBy = get(userSelectedColorByAtom);
 
     if (userSelectedColorBy === null) {
         return VfpParam.THP
     }
+    if (vfpTable?.vfp_type == VfpType_api.VFPINJ && ["WFR", "GFR", "ALQ"].includes(userSelectedColorBy)) {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/settings/atoms/derivedAtoms.ts:167]: Use strict equality ===

**Review by jorgenherje — CHANGES_REQUESTED** [review]
Some comments, otherwise it looks good to me!

Suggest to create an utility function to check if a table is a prod table, as mentioned in comments below. This utility can be updated in the future, s.t. all checks within the module looks to the same "criteria" for categorizing a table as a prod table.

File: backend_py/primary/primary/services/sumo_access/vfp_access.py (hunk 13 above)
Comment (lindjoha) [reply backend_py/primary/primary/services/sumo_access/vfp_access.py:116]: This is python :D

Comment (jorgenherje) [reply backend_py/primary/primary/services/sumo_access/vfp_access.py:116]: Haha. Sorry, my bad! 👍

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 15]
```diff
@@ -1,103 +1,112 @@
...
+            return this._vfpTable.thpUnit
         } 
-        if ("wfr_values" in this._vfpTable) { // This means that it is a VFPPROD table    
-            if (vfpParam == VfpParam.WFR) {
-                return this._vfpTable.wfr_unit
-            } else if (vfpParam == VfpParam.GFR) {
-                return this._vfpTable.gfr_unit
-            } else if (vfpParam == VfpParam.ALQ) {
-                return this._vfpTable.alq_unit
+        if ("isProdTable" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:83]: Use `isProdTable()`-function instead?

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 16]
```diff
@@ -1,103 +1,112 @@
...
             label = "THP"
         } 
-        if ("wfr_values" in this._vfpTable) { // This means that it is a VFPPROD table
-            if (vfpParam == VfpParam.WFR) {
-                label = this._vfpTable.wfr_type
-            } else if (vfpParam == VfpParam.GFR) {
-                label = this._vfpTable.gfr_type
-            } else if (vfpParam == VfpParam.ALQ) {
-                if (this._vfpTable.alq_type === ALQ_api._) {
+        if ("isProdTable" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:100]: Use `isProdTable()`-function instead?

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 17]
```diff
@@ -109,60 +118,64 @@ export class VfpDataAccessor {
     }
 
     getWfrType(): WFR_api {
-        if ("wfr_type" in this._vfpTable) { // This means that it is a VFPPROD table
-            return this._vfpTable.wfr_type
+        if ("isProdTable" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:121]: Use `isProdTable()`-function instead?

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 18]
```diff
@@ -1,103 +1,112 @@
...
             }
         }
-        // throw Error?
-        return ""
+        throw Error("Unhandled vfp parameter ${vfpParam}")
     }
 
     getVfpParamLabel(vfpParam: VfpParam, includeUnit: boolean): string {
         let label = ""
         if (vfpParam == VfpParam.THP) {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:97]: Use strict equality `===`

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 19]
```diff
@@ -109,60 +118,64 @@ export class VfpDataAccessor {
...
+        if ("isProdTable" in this._vfpTable) { // This means that it is a VFPPROD table
+            return this._vfpTable.wfrType
         }
         throw Error("WFR type is not valid for VFPINJ tables.")
     }
 
     getGfrType(): GFR_api {
-        if ("gfr_type" in this._vfpTable) { // This means that it is a VFPPROD table
-            return this._vfpTable.gfr_type
+        if ("isProdTable" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:128]: Use `isProdTable()`-function instead?

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 20]
```diff
@@ -109,60 +118,64 @@ export class VfpDataAccessor {
...
+            return this._vfpTable.gfrType
         }
         throw Error("GFR type is not valid for VFPINJ tables.")
     }
 
     getAlqType(): ALQ_api {
-        if ("alq_type" in this._vfpTable) { // This means that it is a VFPPROD table
-            return this._vfpTable.alq_type
+        if ("isProdTable" in this._vfpTable) { // This means that it is a VFPPROD table
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:135]: Use `isProdTable()`-function instead?

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 21]
```diff
@@ -109,60 +118,64 @@ export class VfpDataAccessor {
...
 
     getVfpProdBhpValues(thpIndex: number, wfrIndex: number, gfrIndex: number, alqIndex: number) : number[] {
-        if ("wfr_values" in this._vfpTable) { // This means that it is a VFPPROD table
-            const nbWfrValues = this._vfpTable.wfr_values.length
-            const nbGfrValues = this._vfpTable.gfr_values.length
-            const nbAlqValues = this._vfpTable.alq_values.length
-            const nbFlowRates = this._vfpTable.flow_rate_values.length
-            const startIndex = nbFlowRates*(nbAlqValues*(nbGfrValues*(nbWfrValues*thpIndex+wfrIndex)+gfrIndex)+alqIndex)
-            return this._vfpTable.bhp_values.slice(startIndex, startIndex+nbFlowRates)
+        if (!("isProdTable" in this._vfpTable)) {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:142]: Use `isProdTable()`-function instead?

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 22]
```diff
@@ -109,60 +118,64 @@ export class VfpDataAccessor {
...
+        const startIndex = nbFlowRates*(nbAlqValues*(nbGfrValues*(nbWfrValues*thpIndex+wfrIndex)+gfrIndex)+alqIndex)
+        return this._vfpTable.bhpValues.slice(startIndex, startIndex+nbFlowRates)
     }
 
     getVfpInjBhpValues(thpIndex: number) : number[] {
-        if (this._vfpTable.vfp_type === VfpType_api.VFPINJ) {
-            const nbFlowRates = this._vfpTable.flow_rate_values.length
-            const startIndex = nbFlowRates*thpIndex
-            return this._vfpTable.bhp_values.slice(startIndex, startIndex+nbFlowRates)
+        if(!("isInjTable" in this._vfpTable)) {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:154]: Use `isProdTable()`-function instead?

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 23]
```diff
@@ -109,60 +118,64 @@ export class VfpDataAccessor {
...
+            } else if (vfpParam === VfpParam.ALQ) {
+                return this._vfpTable.alqValues.length
             }            
         }
-        return NaN
+        throw Error("Unhandled vfp parameter ${vfpParam}")
     }
 }
+
+export function isProdTable(vfpTable: VfpProdTable_api | VfpInjTable_api): vfpTable is VfpProdTable_api {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:179]: Consider placing the function in a separate file in the `utils`-folder. Name it e.g. `vfpTableClassifier.ts`

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:1]: Minor issue: Did not see this last time, but file names should be starting with lower case letters, i.e. `vfpDataAccessor.ts`

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 24]
```diff
@@ -1,103 +1,112 @@
...
     }
 
     getFlowRateLabel(): string {
-        const flowRateType = this._vfpTable.flow_rate_type
+        const flowRateType = this._vfpTable.flowRateType
         const flowRateUnit = this.getFlowRateUnit()
-        if (flowRateType == FlowRateType_api.OIL) {
+        if (flowRateType === FlowRateType_api.OIL) {
             return `Oil Rate (${flowRateUnit})`
         } else if (flowRateType == FlowRateType_api.GAS) {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:35]: Use strict equality `===`

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 25]
```diff
@@ -1,103 +1,112 @@
...
-            } else if (vfpParam == VfpParam.ALQ) {
-                return this._vfpTable.alq_values
+        if ("isProdTable" in this._vfpTable) {
+            if (vfpParam === VfpParam.WFR) {
+                return this._vfpTable.wfrValues
+            } 
+            if (vfpParam === VfpParam.GFR) {
+                return this._vfpTable.gfrValues
+            } 
+            if (vfpParam == VfpParam.ALQ) {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:72]: Use strict equality `===`

File: frontend/src/modules/Vfp/utils/VfpDataAccessor.ts [hunk 26]
```diff
@@ -109,60 +118,64 @@ export class VfpDataAccessor {
...
+            throw Error("The getVfpInjBhpValues function is only valid for Injection VFP tables.")    
         }
-        throw Error("The getVfpInjBhpValues function is not valid for VFPPROD tables.")    
+        const nbFlowRates = this._vfpTable.flowRateValues.length
+        const startIndex = nbFlowRates*thpIndex
+        return this._vfpTable.bhpValues.slice(startIndex, startIndex+nbFlowRates)
     }
 
     getNumberOfValues(vfpParam: VfpParam): number {
         if (vfpParam == VfpParam.THP) {
```
Comment (jorgenherje) [inline frontend/src/modules/Vfp/utils/VfpDataAccessor.ts:163]: Use strict equality `===`

**Review by jorgenherje — CHANGES_REQUESTED** [review]
Looks good to me! Just some minor comments, then it's good to go!

**Pull request merged**